lib/util/Ordering.ts

```diff
@@ -79,49 +77,38 @@ export function orderTypes(termA: RDF.Term | undefined, termB: RDF.Term | undefi
 
 function orderLiteralTypes(litA: RDF.Literal, litB: RDF.Literal,
   typeDiscoveryCallback?: SuperTypeCallback, typeCache?: TypeCache): -1 | 0 | 1 {
-  const defaultTimezone: ITimeZoneRepresentation = { zoneHours: 0, zoneMinutes: 0 };
-
-  const openWorldType: ISuperTypeProvider = {
-    discoverer: typeDiscoveryCallback || (() => 'term'),
-    cache: typeCache || new LRUCache(),
+  const isLess = regularFunctions[C.RegularOperator.LT];
+  const isEqual = regularFunctions[C.RegularOperator.EQUAL];
+  const context = {
+    now: new Date(),
+    functionArgumentsCache: {},
+    superTypeProvider: {
+      discoverer: typeDiscoveryCallback || (() => 'term'),
+      cache: typeCache || new LRUCache(),
+    },
+    defaultTimeZone: { zoneHours: 0, zoneMinutes: 0 },
   };
-  const termTransformer = new TermTransformer(openWorldType);
+
+  const termTransformer = new TermTransformer(context.superTypeProvider);
   const myLitA = termTransformer.transformLiteral(litA);
   const myLitB = termTransformer.transformLiteral(litB);
-  const typeA = myLitA.dataType;
-  const typeB = myLitB.dataType;
-
-  const superTypeDictA: GeneralSuperTypeDict = getSuperTypeDict(typeA, openWorldType);
-  const superTypeDictB: GeneralSuperTypeDict = getSuperTypeDict(typeB, openWorldType);
 
-  // Special handling of specific datatypes
-  if (!isNonLexicalLiteral(myLitA) && !isNonLexicalLiteral(myLitB)) {
-    if (TypeURL.XSD_BOOLEAN in superTypeDictA && TypeURL.XSD_BOOLEAN in superTypeDictB ||
-      TypeAlias.SPARQL_NUMERIC in superTypeDictA && TypeAlias.SPARQL_NUMERIC in superTypeDictB ||
-      TypeURL.XSD_STRING in superTypeDictA && TypeURL.XSD_STRING in superTypeDictB) {
-      return comparePrimitives(myLitA.typedValue, myLitB.typedValue);
+  try {
+    if ((<E.BooleanLiteral> isEqual.apply([ myLitA, myLitB ], context)).typedValue) {
+      return 0;
     }
-    if (TypeURL.XSD_DATE_TIME in superTypeDictA && TypeURL.XSD_DATE_TIME in superTypeDictB) {
-      return comparePrimitives(
-        toUTCDate(myLitA.typedValue, defaultTimezone).getTime(),
-        toUTCDate(myLitB.typedValue, defaultTimezone).getTime(),
-      );
+    if ((<E.BooleanLiteral> isLess.apply([ myLitA, myLitB ], context)).typedValue) {
+      return -1;
     }
-    if (TypeURL.RDF_LANG_STRING in superTypeDictA && TypeURL.RDF_LANG_STRING in superTypeDictB) {
-      const compareType = comparePrimitives(myLitA.typedValue, myLitB.typedValue);
-      if (compareType !== 0) {
-        return compareType;
-      }
-      return comparePrimitives((<LangStringLiteral>myLitA).language, (<LangStringLiteral>myLitB).language);
+    return 1;
```

Comment:
This else case might be false. While implementing https://github.com/comunica/sparqlee/pull/181 I saw that there are cases where `A < B` = False = `A > B` = false = `A < B`. In ordering, it might not be too bad(?) since in this case, we just want consistent ordering? However, this does matter regarding our discussing in https://github.com/comunica/sparqlee/pull/181

Example of such a case is `A = B = NaN`: https://www.w3.org/TR/xpath-functions/#func-numeric-less-than

Reply:
I don't get what the problem is, can you elaborate?

Reply:
Implemented here is a fall through chain:
1. A == B ?
2. A  <  B ?
3. A  >  B

What I mean is that the conclusion in step 3 is "false", since there are cases where non of the above hold. (I think 2 times the value NaN is such a case.) (I'd need to check again, this command is quite old)

Reply:
If this is only the case for NaN, then this shouldn't be applicable, since NaN can't occur in SPARQL AFAIK.
So maybe it's a problem that can never occur?

Reply:
I checked; It seems like [float](https://www.w3.org/TR/2004/REC-xmlschema-2-20041028/#float) does have "NaN" in its value space. More particular, the spec states:

> This datatype differs from that of [[IEEE 754-1985]](https://www.w3.org/TR/2004/REC-xmlschema-2-20041028/#ieee754) in that there is only one NaN and only one zero. This makes the equality and ordering of values in the data space differ from that of [[IEEE 754-1985]](https://www.w3.org/TR/2004/REC-xmlschema-2-20041028/#ieee754) only in that for schema purposes NaN = NaN.

However, in JavaScript, (NaN === NaN) = true, if I'm not mistaken, in current sparqlee implementation, we 
 do thus give a wrong answer.